{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%\n",
    "from cbio_py import cbio_mod as cb\n",
    "\n",
    "studies = cb.getAllStudies()\n",
    "print(f\"Total number of studies: {len(studies)}\")\n",
    "\n",
    "# get the 5 study IDs with the most samples\n",
    "sorted_samples = sorted(studies, key=lambda x: x['allSampleCount'], reverse=True)\n",
    "print(\"Top 5 studies with the most samples:\")\n",
    "for studies in sorted_samples[:5]:\n",
    "    print(f\"Study ID: {studies['studyId']}, Sample Count: {studies['allSampleCount']}, Cancer Type: {studies['cancerTypeId']}\")\n",
    "    \n",
    "print(\"Overall sample count across all studies:\", sum(int(study['allSampleCount']) for study in sorted_samples[:5]))\n",
    "\n",
    "# %%\n",
    "col_studies = []\n",
    "count = 0\n",
    "for study in sorted_samples:\n",
    "    if study['cancerTypeId'] in ['coad', 'read', 'coadread']:\n",
    "        print(f\"Study ID: {study['studyId']}, Sample Count: {study['allSampleCount']}, Cancer Type: {study['cancerTypeId']}\")\n",
    "        count += 1\n",
    "        col_studies.append(study)\n",
    "        \n",
    "print(f\"Total number of studies with 'coad', 'read', or 'coadread' cancer types: {count}\")\n",
    "print(f\"Total sample count for these studies: {sum(int(study['allSampleCount']) for study in col_studies)}\")\n",
    "\n",
    "# %%\n",
    "from bravado.client import SwaggerClient\n",
    "cbioportal = SwaggerClient.from_url('https://www.cbioportal.org/api/v2/api-docs',\n",
    "                                config={\"validate_requests\":False,\"validate_responses\":False,\"validate_swagger_spec\":False})\n",
    "\n",
    "gene_mut = []\n",
    "total_mutations = 0\n",
    "\n",
    "for study in sorted_samples[:5]:\n",
    "    # get the mutations for the first sample in each of the top 5 studies\n",
    "    sample_list = cbioportal.Mutations.getMutationsInMolecularProfileBySampleListIdUsingGET(\n",
    "        molecularProfileId=f\"{study['studyId']}_mutations\",\n",
    "        sampleListId=f\"{study['studyId']}_all\"\n",
    "    ).result()\n",
    "    print(len(sample_list), \"mutations found for study:\", study['studyId'])\n",
    "    total_mutations += len(sample_list)\n",
    "    gene_mut.extend(sample_list)\n",
    "    \n",
    "# for study in col_studies:\n",
    "#     # get the mutations for the first sample in each of the 'coad', 'read', or 'coadread' studies\n",
    "#     sample_list = cbioportal.Mutations.getMutationsInMolecularProfileBySampleListIdUsingGET(\n",
    "#         molecularProfileId=f\"{study['studyId']}_mutations\",\n",
    "#         sampleListId=f\"{study['studyId']}_all\"\n",
    "#     ).result()\n",
    "#     print(len(sample_list), \"mutations found for study:\", study['studyId'])\n",
    "#     total_mutations += len(sample_list)\n",
    "#     gene_mut.extend(sample_list)\n",
    "    \n",
    "print(\"Total mutations considered:\", total_mutations)\n",
    "\n",
    "# %%\n",
    "gene_list = cb.getAllGenes(return_type = 'dict')\n",
    "gene_dict = {gene['entrezGeneId']: gene for gene in gene_list}\n",
    "\n",
    "# %%\n",
    "# output the mutations to a file\n",
    "gene_set = set()\n",
    "with open('../data/top_5_study_mutations.txt', 'w') as f:\n",
    "    for mutation in gene_mut:\n",
    "        if not mutation.entrezGeneId:\n",
    "            print(mutation)\n",
    "        gene = gene_dict[mutation.entrezGeneId]['hugoGeneSymbol']\n",
    "        sample_id = mutation.sampleId\n",
    "        f.write(f\"{gene}\\t{sample_id}\\n\")\n",
    "        gene_set.add(mutation.entrezGeneId)\n",
    "\n",
    "# output the unique genes to a file\n",
    "with open('../data/unique_genes_mutated.txt', 'w') as f:\n",
    "    for gene in gene_set:\n",
    "        f.write(f\"{gene_dict[gene]['hugoGeneSymbol']}\\n\")\n",
    "\n",
    "# %%\n",
    "# create a dataframe of the mutations\n",
    "import pandas as pd\n",
    "mut_df = pd.read_csv('../data/top_5_study_mutations.txt', sep='\\t', header=None, names=['Gene', 'Sample'])\n",
    "mut_df\n",
    "\n",
    "# %%\n",
    "gene_mut_dict = {}\n",
    "for i in range(len(mut_df)):\n",
    "    if mut_df.iloc[i]['Sample'] not in gene_mut_dict:\n",
    "        gene_mut_dict[mut_df.iloc[i]['Sample']] = set()\n",
    "    gene_mut_dict[mut_df.iloc[i]['Sample']].add(mut_df.iloc[i]['Gene'])\n",
    "    \n",
    "gene_mut_dict\n",
    "\n",
    "# %%\n",
    "# Extract unique genes and samples\n",
    "samples = list(gene_mut_dict.keys())\n",
    "genes = sorted(set(g for genes in gene_mut_dict.values() for g in genes))\n",
    "\n",
    "# Create a DataFrame initialized with zeros\n",
    "df = pd.DataFrame(0, index=samples, columns=genes)\n",
    "\n",
    "# Populate the DataFrame\n",
    "for sample, mutated_genes in gene_mut_dict.items():\n",
    "    df.loc[sample, list(mutated_genes)] = 1\n",
    "\n",
    "df\n",
    "\n",
    "# %%\n",
    "# data analysis\n",
    "# top 20 genes with the most mutations\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "top_genes = df.sum().sort_values(ascending=False).head(20)\n",
    "\n",
    "# in the plot, we want to show the number of mutations for each gene\n",
    "top_genes.plot(kind='barh', figsize=(10, 6))\n",
    "plt.title('Top 20 Genes with the Most Mutations')\n",
    "plt.xlabel('Number of Mutations')\n",
    "# in each bar, we want to show the gene count/total samples\n",
    "for index, value in enumerate(top_genes):\n",
    "    plt.text(value, index, str(value), va='center')\n",
    "plt.xticks(rotation=45)\n",
    "plt.gca().invert_yaxis()  # invert y axis to have the highest count on top\n",
    "plt.grid(axis='x', linestyle='--', alpha=0.7)\n",
    "plt.xlabel('Number of Mutations')\n",
    "plt.ylabel('Gene')\n",
    "plt.tight_layout()\n",
    "plt.savefig('../data/top_20_genes_mutations.png')\n",
    "plt.show()\n",
    "\n",
    "\n",
    "# %%\n",
    "# save df to a file\n",
    "df.to_csv('../data/mutation_matrix.csv', index=True)\n",
    "\n",
    "# %%\n",
    "sample_cd = {}\n",
    "\n",
    "for study in sorted_samples[:5]:\n",
    "    print(f\"Getting clinical data for study: {study['studyId']}\")\n",
    "    data = cb.getAllClinicalDataInStudy(studyId=study['studyId'])\n",
    "    print(f\"Number of clinical data samples in study {study['studyId']}: {len(data)}\")\n",
    "    \n",
    "    for sample in data:\n",
    "        if sample['sampleId'] not in sample_cd:\n",
    "            sample_cd[sample['sampleId']] = []\n",
    "            \n",
    "        sample_cd[sample['sampleId']].append(sample)\n",
    "        \n",
    "    print(f\"Total number of clinical data for now: {len(sample_cd)}\")\n",
    "        \n",
    "# write the clinical data to a file\n",
    "with open('../data/clinical_data.txt', 'w') as f:\n",
    "    for sample, data in sample_cd.items():\n",
    "        for d in data:\n",
    "            f.write(f\"{sample}\\t{d['clinicalAttributeId']}\\t{d['studyId']}\\t{d['value']}\\n\")\n",
    "\n",
    "# %%\n",
    "mutation_matrix = pd.read_csv('../data/mutation_matrix.csv')\n",
    "for sampleid in mutation_matrix['Sample_ID'].values.tolist():\n",
    "    if sampleid not in sample_cd:\n",
    "        print(f\"Sample ID {sampleid} not found in clinical data for the top 5 studies.\")\n",
    "        \n",
    "# extend 2 columns to the mutation matrix: msi, stage, initialize with nan\n",
    "mutation_matrix['MSI'] = float('nan')\n",
    "mutation_matrix['Stage'] = float('nan')\n",
    "mutation_matrix\n",
    "\n",
    "# %%\n",
    "total_case = len(mutation_matrix)\n",
    "msi_count = 0\n",
    "stage_count = 0\n",
    "\n",
    "for id in mutation_matrix['Sample_ID'].values.tolist():\n",
    "    for cd_data in sample_cd[id]:\n",
    "        if cd_data['clinicalAttributeId'] == 'PATHOLOGICAL_GROUP' or cd_data['clinicalAttributeId'] == 'CLINICAL_STAGE':\n",
    "            if cd_data['clinicalAttributeId'] == 'PATHOLOGICAL_GROUP':\n",
    "            # for pathological group, we want to use the value as stage\n",
    "                mutation_matrix.loc[mutation_matrix['Sample_ID'] == id, 'Stage'] = cd_data['value']\n",
    "            elif cd_data['clinicalAttributeId'] == 'CLINICAL_STAGE':\n",
    "                mutation_matrix.loc[mutation_matrix['Sample_ID'] == id, 'Stage'] = cd_data['value']\n",
    "            stage_count += 1\n",
    "            \n",
    "        if cd_data['clinicalAttributeId'] == 'MSI_TYPE':\n",
    "            mutation_matrix.loc[mutation_matrix['Sample_ID'] == id, 'MSI'] = cd_data['value']\n",
    "            msi_count += 1\n",
    "\n",
    "print(f\"Total samples with stage data: {stage_count}/{total_case}\")\n",
    "print(f\"Total samples with msi data: {msi_count}/{total_case}\")\n",
    "        \n",
    "# # sampleId=P-0001534-T01-IM3&studyId=msk_chord_2024\n",
    "# cb.getAllClinicalDataOfSampleInStudy(studyId='msk_chord_2024', sampleId='P-0001534-T01-IM3')\n",
    "\n",
    "# %%\n",
    "mutation_matrix.to_csv('../data/mutation_matrix_with_clinical_data.csv', index=False)\n",
    "\n",
    "# %%\n",
    "from bravado.client import SwaggerClient\n",
    "cbioportal = SwaggerClient.from_url('https://www.cbioportal.org/api/v2/api-docs',\n",
    "                                config={\"validate_requests\":False,\"validate_responses\":False,\"validate_swagger_spec\":False})\n",
    "\n",
    "gene_mut = []\n",
    "total_mutations = 0\n",
    "\n",
    "\n",
    "# get the mutations for the first sample in each of the top 5 studies\n",
    "sample_list = cbioportal.Mutations.getMutationsInMolecularProfileBySampleListIdUsingGET(\n",
    "    molecularProfileId=f\"msk_chord_2024_mutations\",\n",
    "    sampleListId=f\"msk_chord_2024_all\"\n",
    ").result()\n",
    "print(len(sample_list), \"mutations found for study:\", 'msk_chord_2024')\n",
    "total_mutations += len(sample_list)\n",
    "gene_mut.extend(sample_list)\n",
    "    \n",
    "print(\"Total mutations considered:\", total_mutations)\n",
    "\n",
    "# %%\n",
    "gene_mut[0]\n",
    "\n",
    "# %%\n",
    "gene_list = cb.getAllGenes(return_type = 'dict')\n",
    "gene_dict = {gene['entrezGeneId']: gene for gene in gene_list}\n",
    "\n",
    "# output the mutations to a file\n",
    "gene_set = set()\n",
    "with open('../data/msk_2024_mutations.txt', 'w') as f:\n",
    "    f.write(\"Gene\\tSample\\tPatient\\tmutationType\\tvariantType\\tproteinChange\\tchr\\tstartPosition\\tendPosition\\treferenceAllele\\tvariantAllele\\tproteinPosStart\\tproteinPosEnd\\tmutationStatus\\ttumorAltCount\\ttumorRefCount\\tnormalAltCount\\tnormalRefCount\\n\")\n",
    "    for mutation in gene_mut:\n",
    "        if not mutation.entrezGeneId:\n",
    "            print(mutation)\n",
    "        gene = gene_dict[mutation.entrezGeneId]['hugoGeneSymbol']\n",
    "        sample_id = mutation.sampleId\n",
    "        f.write(f\"{gene}\\t{sample_id}\\t{mutation.patientId}\\t{mutation.mutationType}\\t{mutation.variantType}\\t{mutation.proteinChange}\\t{mutation.chr}\\t{mutation.startPosition}\\t{mutation.endPosition}\\t{mutation.referenceAllele}\\t{mutation.variantAllele}\\t{mutation.proteinPosStart}\\t{mutation.proteinPosEnd}\\t{mutation.mutationStatus}\\t{mutation.tumorAltCount}\\t{mutation.tumorRefCount}\\t{mutation.normalAltCount}\\t{mutation.normalRefCount}\\n\")\n",
    "        gene_set.add(mutation.entrezGeneId)\n",
    "\n",
    "# %%\n",
    "data = cb.getAllClinicalDataInStudy(studyId='msk_chord_2024')\n",
    "data[:50]\n",
    "\n",
    "\n",
    "# %%\n",
    "data_1 = cb.getClinicalAttributesByStudyId(studyId='msk_chord_2024')\n",
    "data_1\n",
    "\n",
    "# %%\n",
    "data_2 = cb.getAllPatientsInStudy(studyId='msk_chord_2024')\n",
    "data_2\n",
    "\n",
    "# %%\n",
    "data3 = cb.getClinicalAttributeInStudy(studyId='msk_chord_2024', clinicalAttributeId='OS_MONTHS')\n",
    "data3\n",
    "\n",
    "# %%\n",
    "data_4 = cb.getAllClinicalDataOfPatientInStudy(studyId='msk_chord_2024', patientId='P-0002773')\n",
    "data_4\n",
    "\n",
    "# %%\n",
    "import pandas as pd\n",
    "\n",
    "df = pd.read_csv('../data/msk_2024_mutations.txt', sep='\\t')\n",
    "df.head()\n",
    "\n",
    "print(len(df['Sample'].unique()), \"unique samples in the mutation data\")\n",
    "print(len(df['Patient'].unique()), \"unique patients in the mutation data\")\n",
    "\n",
    "# how many patient has more than 1 sample\n",
    "patient_sample_count = df.groupby('Patient')['Sample'].nunique()\n",
    "print(\"Number of patients with more than 1 sample:\")\n",
    "print(patient_sample_count[patient_sample_count > 1].count())\n",
    "print(\"Number of patients with only 1 sample:\")\n",
    "print(patient_sample_count[patient_sample_count == 1].count())\n",
    "print(\"Number of patients with less 1 sample:\")\n",
    "print(patient_sample_count[patient_sample_count < 1].count())\n",
    "\n",
    "# exclude the patients with more than 1 sample\n",
    "df_single = df[df['Patient'].isin(patient_sample_count[patient_sample_count == 1].index)]\n",
    "print(len(df_single['Sample'].unique()), \"unique samples in the mutation data after excluding patients with more than 1 sample\")\n",
    "print(len(df_single['Patient'].unique()), \"unique patients in the mutation data after excluding patients with more than 1 sample\")\n",
    "\n",
    "# %%\n",
    "print(f\"Mutation Type: {df_single['mutationType'].unique()} ({len(df_single['mutationType'].unique())})\")   \n",
    "print(f\"Variant Type: {df_single['variantType'].unique()} ({len(df_single['variantType'].unique())})\")\n",
    "print(f\"Protein Change: {df_single['proteinChange'].unique()} ({len(df_single['proteinChange'].unique())})\")\n",
    "print(f\"Chromosome: {df_single['chr'].unique()} ({len(df_single['chr'].unique())})\")\n",
    "print(f\"Start Position: {df_single['startPosition'].unique()} ({len(df_single['startPosition'].unique())})\")\n",
    "print(f\"End Position: {df_single['endPosition'].unique()} ({len(df_single['endPosition'].unique())})\")\n",
    "print(f\"Reference Allele: {df_single['referenceAllele'].unique()} ({len(df_single['referenceAllele'].unique())})\")\n",
    "print(f\"Variant Allele: {df_single['variantAllele'].unique()} ({len(df_single['variantAllele'].unique())})\")\n",
    "print(f\"Protein Position Start: {df_single['proteinPosStart'].unique()} ({len(df_single['proteinPosStart'].unique())})\")\n",
    "print(f\"Protein Position End: {df_single['proteinPosEnd'].unique()} ({len(df_single['proteinPosEnd'].unique())})\")\n",
    "print(f\"Mutation Status: {df_single['mutationStatus'].unique()} ({len(df_single['mutationStatus'].unique())})\")\n",
    "print(f\"Tumor Alt Count: {df_single['tumorAltCount'].unique()} ({len(df_single['tumorAltCount'].unique())})\")\n",
    "print(f\"Tumor Ref Count: {df_single['tumorRefCount'].unique()} ({len(df_single['tumorRefCount'].unique())})\")\n",
    "print(f\"Normal Alt Count: {df_single['normalAltCount'].unique()} ({len(df_single['normalAltCount'].unique())})\")\n",
    "print(f\"Normal Ref Count: {df_single['normalRefCount'].unique()} ({len(df_single['normalRefCount'].unique())})\")\n",
    "\n",
    "df_single\n",
    "\n",
    "# %%\n",
    "# select mutationType, variantType, chr, startPosition, endPosition, proteinPosStart, proteinPosEnd, tumorAlleleFreq = tumorAltCount/(tumorAltCount + tumorRefCount), normalAlleleFreq = normalAltCount/(normalAltCount + normalRefCount)\n",
    "df_single['tumorAlleleFreq'] = df_single['tumorAltCount'] / (df_single['tumorAltCount'] + df_single['tumorRefCount'])\n",
    "df_single['normalAlleleFreq'] = df_single['normalAltCount'] / (df_single['normalAltCount'] + df_single['normalRefCount'])\n",
    "\n",
    "df_clean = df_single[['Gene', 'Sample', 'Patient', 'mutationType', 'variantType', 'chr', 'startPosition', 'endPosition',\n",
    "                       'proteinPosStart', 'proteinPosEnd', 'tumorAlleleFreq', 'normalAlleleFreq']]\n",
    "df_clean.to_csv('../data/msk_2024_mutations_clean.csv', index=False)\n",
    "df_clean.head()\n",
    "\n",
    "# %%\n",
    "df['chr'] = df['chr'].astype(str)  # ensure string type if numeric chromosome\n",
    "categorical_cols = ['mutationType', 'variantType', 'chr']\n",
    "df_encoded = pd.get_dummies(df, columns=categorical_cols, drop_first=False)\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "dlg",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "name": "python",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
